mst/mst.go

```diff
@@ -346,109 +598,105 @@ func (mst *MerkleSearchTree) Delete(ctx context.Context, k string) (*MerkleSearc
+	copy(nents, entries[:ix])
+	nents[ix] = entry
+	copy(nents[ix+1:], entries[ix+1:])
+
+	checkTreeInvariant(nents)
+	return mst.newTree(nents), nil
+}
+
+// "remove entry at index"
+func (mst *MerkleSearchTree) removeEntry(ctx context.Context, ix int) (*MerkleSearchTree, error) {
+	entries, err := mst.getEntries(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	if mylayer != olayer {
-		return nil, fmt.Errorf("trying to merge two nodes from different layers")
-	}
+	nents := make([]NodeEntry, len(entries)-1)
+	copy(nents, entries[:ix])
+	copy(nents[ix:], entries[ix+1:])
+
+	checkTreeInvariant(nents)
+	return mst.newTree(nents), nil
+}
 
+// "append entry to end of the node"
+func (mst *MerkleSearchTree) append(ctx context.Context, ent NodeEntry) (*MerkleSearchTree, error) {
 	entries, err := mst.getEntries(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	tomergeEnts, err := omst.getEntries(ctx)
+	nents := make([]NodeEntry, len(entries)+1)
+	copy(nents, entries)
+	nents[len(nents)-1] = ent
+
+	checkTreeInvariant(nents)
+	return mst.newTree(nents), nil
+}
+
+// "prepend entry to start of the node"
+func (mst *MerkleSearchTree) prepend(ctx context.Context, ent NodeEntry) (*MerkleSearchTree, error) {
+	entries, err := mst.getEntries(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	lastInLeft := entries[len(entries)-1]
-	firstInRight := entries[0]
-
-	if lastInLeft.isTree() && firstInRight.isTree() {
-		merged, err := lastInLeft.Tree.appendMerge(ctx, firstInRight.Tree)
-		if err != nil {
-			return nil, err
-		}
+	nents := make([]NodeEntry, len(entries)+1)
+	copy(nents[1:], entries)
+	nents[0] = ent
 
-		return mst.newTree(append(append(entries[:len(entries)-1], treeEntry(merged)), tomergeEnts[1:]...)), nil
-	} else {
-		return mst.newTree(append(entries, tomergeEnts...)), nil
-	}
+	checkTreeInvariant(nents)
+	return mst.newTree(nents), nil
 }
 
-var ErrNotFound = fmt.Errorf("mst: not found")
-
-func (mst *MerkleSearchTree) Get(ctx context.Context, k string) (cid.Cid, error) {
-	index, err := mst.findGtOrEqualLeafIndex(ctx, k)
-	if err != nil {
-		return cid.Undef, err
-	}
-
-	found, err := mst.atIndex(index)
-	if err != nil {
-		return cid.Undef, err
-	}
-
-	if !found.isUndefined() && found.isLeaf() && found.Key == k {
-		return found.Val, nil
-	}
-
-	prev, err := mst.atIndex(index - 1)
+// "returns entry at index"
+// Apparently returns null if nothing at index, which seems brittle
+func (mst *MerkleSearchTree) atIndex(ix int) (NodeEntry, error) {
+	entries, err := mst.getEntries(context.TODO())
 	if err != nil {
-		return cid.Undef, err
-	}
-
-	if !prev.isUndefined() && prev.isTree() {
-		return prev.Tree.Get(ctx, k)
+		return NodeEntry{}, err
 	}
 
-	return cid.Undef, ErrNotFound
-}
-
-func (mst *MerkleSearchTree) createParent(ctx context.Context) (*MerkleSearchTree, error) {
-	layer, err := mst.getLayer(ctx)
-	if err != nil {
-		return nil, err
+	// TODO(bnewbold): same as Typescript, but shouldn't this error instead of returning null?
```

Comment:
Its a bit weird, but to match typescript the closest ive let the zero value of NodeEntry.Kind be 'undefined', so we can have the same explicitly undefined checks that the typescript code does.
If i was implementing this from scratch and not from referenced i'd definitely use typed errors